src/main/java/org/sagebionetworks/bridge/spring/controllers/CRCController.java

```diff
@@ -0,0 +1,372 @@
+        
+        updateState(account, state);
+        accountService.updateAccount(account, null);
+        return status;
+    }    
+    
+    private void updateState(StudyParticipant.Builder builder, AccountStates state) {
+        StudyParticipant p = builder.build();
+        Set<String> mutableGroups = new HashSet<>(); 
+        mutableGroups.addAll(p.getDataGroups());
+        mutableGroups.removeAll(ALL_STATES);
+        mutableGroups.add(state.name().toLowerCase());
+        builder.withDataGroups(mutableGroups);
+        
+        Map<String, String> atts = new HashMap<>();
+        atts.putAll(p.getAttributes());
+        atts.put(TIMESTAMP_FIELD, getTimestamp().toString());
+        builder.withAttributes(atts);
+    }
+    
+    private void updateState(Account account, AccountStates state) {
+        Set<String> mutableGroups = new HashSet<>(); 
+        mutableGroups.addAll(account.getDataGroups());
+        mutableGroups.removeAll(ALL_STATES);
+        mutableGroups.add(state.name().toLowerCase());
+        account.setDataGroups(mutableGroups);
+
+        account.getAttributes().put(TIMESTAMP_FIELD, getTimestamp().toString());
+    }
+    
+    // This is bound to one specific account that we've given to Columbia. 
+    // No one else can authenticate for these calls, and the account itself
+    // has no administrative roles so it can't do anything else besides 
+    // calling these methods.
+    void httpBasicAuthentication() {
+        String value = request().getHeader(AUTHORIZATION);
+        if (value == null || value.length() < 5) {
+            throw new NotAuthenticatedException();
+        }
+        // Remove "Basic ";
+        value = value.substring(5).trim();
+
+        // Decode the credentials from base 64
+        value = new String(Base64.decodeBase64(value));
+        // Split to username and password
+        String[] credentials = value.split(":");
+        if (credentials.length != 2) {
+            throw new NotAuthenticatedException();
+        }
+        if (!ACCOUNTS.keySet().contains(credentials[0])) {
+            throw new NotAuthenticatedException();
+        }
+        String appId = ACCOUNTS.get(credentials[0]);
+        SignIn signIn = new SignIn.Builder()
+                .withAppId(appId)
+                .withExternalId(credentials[0])
+                .withPassword(credentials[1]).build();
+        App app = appService.getApp(appId);
+        
+        // Verify the password
+        Account account = accountService.authenticate(app, signIn);
+
+        // This method of verification entirely sidesteps RequestContext 
+        // initialization. Set up anything that is needed in this controller.
+        BridgeUtils.setRequestContext(new RequestContext.Builder()
+                .withCallerAppId(appId)
+                .withCallerSubstudies(BridgeUtils.collectSubstudyIds(account))
+                .build());
```

Comment:
We have code here that sets the app ID and substudies into the request context, but this isn't tested anywhere. Should we add a test for this?

Reply:
I feel this is tested by the post* methods, which fail if the setting of RequestContext is removed (either of the two values will cause tests to fail for lack of that value in other methods).